### PR TITLE
Feat/feature flagging/simplify usage

### PR DESCRIPTION
**Pull request by odinr**
# @equinor/fusion-framework-module-feature-flag - major

- removed function `enableCgi`, must be enabled by adding plugin _(see [URL](#url))_
- removed function `enableApi`, must be enabled by adding plugin _(see [Local](#local))_
- added ability to provide order _(`int`)_ to the plugin

  - initial values will be merged in the order which plugins are sorted by

- added getter for current features
- added function for providing callback handler of `onFeatureToggle`
- simplified `getFeature`
  - now only takes key of feature for selection
  - will now return the feature flag from current state
  - allow value type hint in template
- simplified `getFeatures`
  - will now return the feature flags from current state
- added `EventModule`
  - `onFeatureFlagToggle` - before feature is toggled _(cancelable)_
  - `onFeatureFlagsToggled` - after features are toggled _(state has changed)_

> [!CAUTION] 
> `getFeature` and `getFeatures` no longer returns a stream, but current state.
> See [Selectors](#selectors)

removed plugin CGI, since confusing naming, see [URL](#url)

```ts
featureConfigurator.addPlugin(
  /** array of IFeatureFlag */
  createUrlPlugin(flags),
);
```

> [!WARNING]
> the new url plugin does not have any storage, see [Local](#local)

Added a plugin for handling local flags _(store toggle state to session or local storage)_

```ts
featureConfigurator.addPlugin(
  createLocalStoragePlugin(flags, {
    name: `uniq-name`,
    // type: 'local' | 'local'
  }),
);
```

> [!TIP]
> session storage kan be enabled by options

renamed and refactored feature flag control by URL

> [!IMPORTANT]
> URL plugin should should only have keys of features to toggle, this plugin is meant in peer with either `API` or `local`

simplified usage by removing required `name`

> [!WARNING]
> the `namespace` must be uniq for each implementation

moved selectors from `FeatureFlagProvider.selector.ts`

```ts
import { findFeature } from "@equinor/fusion-framework-module-feature-flag/selectors";
const feature$ = provider.feature$.pipe(findFeature("my-feature"));
```

```ts
import { filterFeatures } from "@equinor/fusion-framework-module-feature-flag/selectors";
const features$ = provider.feature$.pipe(filterFeatures((x) => x.enabled));
```

# @equinor/fusion-framework-cli - minor

refactor side-sheet for displaying framework and application feature flags

# @equinor/fusion-framework-react - minor

Added utilities for working with app and framework modules

- Created a custom hook for getting a module from the current application
- Created a custom hook for getting a module from the framework

# @equinor/fusion-framework-app - minor

Added method to `AppConfigurator` to enabled `Feature Flag Module`

```ts
export const configure: ModuleInitiator = (appConfigurator, args) => {
  /** provide a list of features that should be available in the application */
  appConfigurator.useFeatureFlags([
    {
      key: MyFeatures.MyFlag,
      title: "this is a flag",
    },
    {
      key: MyFeatures.MyUrlFlag,
      title: "this feature can be toggled by ?my-url-flag=true",
      allowUrl: true,
    },
  ]);
};
```

# @equinor/fusion-framework-react-app - minor

refactor hook for accessing feature flags

> [!IMPORTANT]
> the `useFeature` hook look for flag in parent **(portal)** if not defined in application scope

> [!CAUTION] > `@equinor/fusion-framework-react-app/feature-flag` will only return `useFeature`, since we don not see any scenario which an application would need to access multiple.
> We might add `useFeatures` if the should be an use-case

# @equinor/fusion-framework-react - minor

refactor framework feature flag hooks

- useCurrentAppFeatures
- _useFeature (internal)_
- _useFeatures (internal)_
- useFrameworkFeature
- useFrameworkFeatures

# @equinor/fusion-framework-cookbook-app-react-feature-flag - minor

refactor cookbook

- simplified example
- updated config
- updated components
- added event logger
- shorten app key

### Check off the following:
- [x] Confirm that I checked changes to branch which I am merging into.
  - _I have validated included files_
  - _My code does not generate new linting warnings_
  - _My PR is not a duplicate, [check existing pr`s](https://github.com/equinor/fusion-framework/pulls)_
- [x] Confirm that the I have completed the [self-review checklist](https://github.com/equinor/fusion-framework/blob/main/contributing/self-review.md).

- [x] Confirm that my changes meet our [code of conduct](https://github.com/equinor/fusion-framework/blob/main/CODE_OF_CONDUCT.md).

#### Conditional
- [x] Confirm project selected and category, actual, iteration are set
- [x] Confirm Milestone selected _(if any)_

### CODE REVIEW

File: packages/react/app/src/feature-flag/useFeature.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            }),
+        );
+    }, [appProvider, frameworkProvider]);
+
+    /** find feature flag by the provided key */
+    const feature$ = useMemo(() => features$.pipe(findFeature<T>(key)), [features$, key]);
+
+    const { value: feature, error } = useObservableState(feature$ ?? EMPTY);
+
+    const toggleFeature = useCallback(
```
Comment (eikeland) [inline packages/react/app/src/feature-flag/useFeature.ts:53]: in ffw-react we use `setEnabled` and `toggle`, and here its `toggleFeature` should it be the same?

File: vue-press/src/.vuepress/sidebar.ts [hunk 2]
```diff
@@ -28,62 +28,78 @@ export default sidebar({
...
-                            link: 'react.md',
-                        },
-                    ]
+            ]
 
+        },
+        {
+            text: 'Feature flag',
+            prefix: 'feature-flag/',
+            link: 'feature-flag/',
```
Comment (eikeland) [inline vue-press/src/.vuepress/sidebar.ts:67]: The landing page for modules/feature-flag/ has no content, the landing page should be the child `module`

File: packages/react/app/src/feature-flag/useFeature.ts (hunk 1 above)
Comment (odinr) [reply packages/react/app/src/feature-flag/useFeature.ts:53]: renamed to toggleFeature

File: cookbooks/app-react-feature-flag/tsconfig.json [hunk 3]
```diff
@@ -17,7 +17,7 @@
   ],
   "include": [
     "src/**/*"
-  ],
+, "app.manifest.config.ts"  ],
```
[the remark this answers is not in the document]
Comment (odinr) [reply cookbooks/app-react-feature-flag/tsconfig.json]: ide drag error, fixed

File: vue-press/src/.vuepress/sidebar.ts (hunk 2 above)
Comment (odinr) [reply vue-press/src/.vuepress/sidebar.ts]: default link for menu item is now module.md

File: packages/react/framework/src/feature-flag/useFeature.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * Custom hook for accessing a feature flag value and toggling it.
+ *
+ * @template T - The type of the feature flag value.
+ * @param {IFeatureFlagProvider} provider - The feature flag provider.
+ * @param {string} key - The key of the feature flag.
+ */
+export const useFeature = <T = unknown>(provider: IFeatureFlagProvider, key: string) => {
+    const feature$ = useMemo(() => provider.features$.pipe(findFeature<T>(key)), [provider, key]);
+    const { value: feature, error } = useObservableState(feature$ ?? EMPTY);
+    const toggle = useCallback(
```
Comment (eikeland) [inline packages/react/framework/src/feature-flag/useFeature.ts:21]: could this be named ``toggleFeature``also?

**Comment by eikeland** [comment]
@equinor/fusion-framework-react-app gives me build errors

File: packages/react/framework/src/feature-flag/useFrameworkFeature.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * @template T - The type of the feature.
+ * @param {string} key - The feature flag key.
+ * @returns {ReturnType<typeof useFeature<T>>} - The result of the feature hook.
+ * @throws {Error} - If feature flagging is not enabled in the framework.
+ */
+export const useFrameworkFeature = <T>(key: string): ReturnType<typeof useFeature<T>> => {
+    const provider = useFrameworkModule<FeatureFlagModule>('featureFlag');
+    if (!provider) {
+        throw Error('Feature flagging is not enabled by in the framework');
```
Comment (eikeland) [inline packages/react/framework/src/feature-flag/useFrameworkFeature.ts:17]: same typo as earlier here.

**Review by eikeland — APPROVED** [review]

**Pull request merged**